[IPC part 4] test cases for IPC strategies.

rpdome opened the pull request:
part 1: #1088
part 2: #1090
part 3: #1092 

- Add test cases that covers the 3 IPC strategies.
    - Also add enum 'Type' in BrokerCommunicationException.
- Add test cases to cover BrokerOperationExecutor
    - Also change the thrown exception type of IIpcStrategy.communicateToBroker() from BaseException to BrokerCommunicationException (make it more specific).
- ~~exclude 'asm' module in org.robolectric:robolectric~~ Issue only shows up in Android Studio 4.1, will figure it out later.
   - nimbus-jose-jwt is using asm 5.0.4, while Robolectric uses 7.0

shahzaibj commented:
The `asm` issue -  did that only happen for you locally or did that also happen on the pipeline (that was run as part of this PR)?

rpdome commented:
> The `asm` issue - did that only happen for you locally or did that also happen on the pipeline (that was run as part of this PR)?

Local (otherwise we should've detected this in my previous PR already. I also ran that test with that commit locally and ran into the same issue).

AdamBJohnsonx commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/mock/MockContextForBoundServiceConnectionFailed.java:
```diff
@@ -0,0 +1,43 @@
+@Implements(BoundServiceClient.class)
+public class MockContextForBoundServiceConnectionFailed<T extends IInterface> {
+    protected @NonNull T connect(@NonNull final String targetServicePackageName)
+            throws ClientException, InterruptedException, TimeoutException, ExecutionException {
+        throw new TimeoutException("Connection timed out");
```
might be nice to parameterize this thing so you could make it throw them all by using a static factory that takes the exception to throw.

rpdome replied:
can you parameterize a shadow class?

AdamBJohnsonx replied:
I was thinking more about whether a shadow is the right way to deal with this, since BoundServiceClient is, I think, our own class.  We shouldn't *need* to use a shadow for it.

rpdome replied:
we don't necessarily need shadow, but IMO it's easier/cleaner to implement this way.

Alternatively, we would have to supply IIpcStrategy object in every methods, across all 3 strategies. (right now each method are calling testOperationXXXXX(), and that function is pulling the strategy object from getStrategy()). Shadowing allows us to easily modify the injected object in specific test cases.

AdamBJohnsonx commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/mock/MockContextForBoundServiceWithSuccessResult.java:
```diff
@@ -0,0 +1,89 @@
+            throws ClientException, InterruptedException, TimeoutException, ExecutionException {
+        final IMicrosoftAuthService authService = new IMicrosoftAuthService() {
+            @Override public IBinder asBinder() {
+                return null;
```
Does it matter what this returns?  If we're really expecting that it isn't called, throw UnsupportedOperationException and then you'll know if that idea changes.  If it is being called, what happenes if it doesn't return null?

rpdome replied:
We don't really use it here. (it should be used if we're doing the actual bind)

AdamBJohnsonx commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/IpcStrategyTests.java:
```diff
@@ -0,0 +1,178 @@
+        return new BrokerOperationBundle(operation, "MOCK_BROKER", bundle);
+    }
+
+    public static Bundle getMockIpcResultBundle() {
```
Do we need to use a fake one?  Could we actually just pull a sample Bundle from the actual code?

rpdome replied:
I intentionally did not use the actual sample bundle - since that depends on the protocol.
This test case should be protocol-agnostic.

AdamBJohnsonx reviewed (commented):
A bunch of small comments.  The enum isn't a concern if we don't send it from server to client, but if it is, flatten it to string or something you can interpret with json.

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/broker/BrokerAccountServiceClient.java:
```diff
@@ -82,14 +85,16 @@ public BrokerAccountServiceClient(@NonNull final Context context,
             default:
-                throw new BaseException("Operation not supported. Wrong service bound.");
+                throw new BrokerCommunicationException(OPERATION_NOT_SUPPORTED_ON_CLIENT_SIDE,
+                        "Operation not supported. Wrong service bound.",
```
It means you're using the wrong BoundServiceClient. will change to Wrong BoundServiceClient used.

rpdome commented on common/src/main/java/com/microsoft/identity/common/exception/BrokerCommunicationException.java:
```diff
@@ -23,19 +23,32 @@
 public class BrokerCommunicationException extends BaseException {
     private static final long serialVersionUID = 4959278068787428329L;
 
+    public enum Type {
+        OPERATION_NOT_SUPPORTED_ON_CLIENT_SIDE,
+        OPERATION_NOT_SUPPORTED_ON_SERVER_SIDE,
+        CONNECTION_ERROR
+    }
```
We're not communicating this to the other side of IPC.

rpdome replied:
Added "StrategyType" into BrokerCommunicationException's argument list (so that it prints more details into the log).

rpdome commented on common/src/main/java/com/microsoft/identity/common/internal/broker/BoundServiceClient.java:
```diff
@@ -127,15 +129,16 @@ public BoundServiceClient(@NonNull final Context context,
-                    BOUND_SERVICE_UNAVAILABLE_OR_NOT_SUPPORTED,
-                    errorMessage);
+            throw new BrokerCommunicationException(
+                    OPERATION_NOT_SUPPORTED_ON_SERVER_SIDE,
```
Not at this point. In this layer we're focusing on Bound Service only (and it failed).

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/brokeroperationexecutor/BrokerOperationExecutorTests.java:
```diff
@@ -0,0 +1,300 @@
+    @Test
+    public void testTwoStrategiesWithTheFirstOneReturningCorruptedResult() {
+        final List<IIpcStrategy> strategyList = new ArrayList<>();
+        strategyList.add(getStrategyWithCorruptedResult());
```
Corrupted here doesn't mean data loss, that, AFAIK, never happens.

It means the broker is sending a result bundle with unexpected format, or the data is not properly compressed. The data is 'corrupted' from the source, therefore it doesn't matter which communication channel you use - you'll always get the same result bundle.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/IpcStrategyTests.java:
```diff
@@ -0,0 +1,178 @@
+import static com.microsoft.identity.common.internal.broker.ipc.BrokerOperationBundle.Operation.MSAL_SIGN_OUT_FROM_SHARED_DEVICE;
+
+/**
+ * IMPORTANT: This class must cover EVERY {@link com.microsoft.identity.common.internal.broker.ipc.BrokerOperationBundle.Operation}.
```
I assume you mean making IIpcStrategy the parameter in the parameterized test.

Each strategy requires different shadow classes.... so ... we could, but I don't want to put all shadow classes in one place.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/IpcStrategyTests.java:
```diff
@@ -0,0 +1,178 @@
+ */
+public abstract class IpcStrategyTests {
+
+    abstract IIpcStrategy getStrategy();
```
Please see the answer regarding parameterized test below.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/IpcStrategyTests.java:
```diff
@@ -0,0 +1,178 @@
+            Assert.fail("Operation should fail.");
+        } catch (BaseException e) {
+            Assert.assertTrue(e instanceof BrokerCommunicationException);
+            Assert.assertSame(((BrokerCommunicationException) e).getType(), CONNECTION_ERROR);
```
same identity?

AdamBJohnsonx replied:
AssertSame is a comparison of object identity, essentially memory address.  So my question is - is it important that it's the same object or should it just need to compare equals?

rpdome replied:
ah ok. assertEquals makes more sense then. not sure why this is not throwing an error :/

AdamBJohnsonx replied:
Because it's an enum. ;-)

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/brokeroperationexecutor/BrokerOperationExecutorTests.java:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    // Providing 2 strategies and the first one fails.
+    // NOTE: This should never happen in real life. broker should return the same values regardless of communication strategy.
```
In this test suite, there are 4 possible outcomes.
1. Gets a bundle back, the bundle contains a valid result. The result returns true.
2. Gets a bundle back, the bundle contains a valid result. The result returns false.
3. Gets a bundle back, the bundle is invalid. (Failure of Strategy)
4. Fails to get the bundle. (Failure to communicate)

I think I used bad wording here. This test case refers to 2)

rpdome replied:
Come to think of it, I should just combine 1 and 2. It doesn't really matter in this layer.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/brokeroperationexecutor/BrokerOperationExecutorTests.java:
```diff
@@ -0,0 +1,300 @@
+        expectCorruptedBundleException(strategyList);
+    }
+
+    // Providing 2 strategies and the last one failed with BrokerCommunicationException inside IIpcStrategy.
```
Yep. This test case is to make sure that the order of strategy matters.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/IpcStrategyTests.java:
```diff
@@ -0,0 +1,178 @@
+
+            if (objA instanceof Bundle && objB instanceof Bundle) {
+                return validateResultBundle((Bundle) objA, (Bundle) objB);
+            } else if (!objA.equals(objB)) {
```
Bundle's equal is not implemented.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/IpcStrategyTests.java:
```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    private boolean validateResultBundle(@NonNull final Bundle resultBundle,
```
added testImplementation dependency to testutils and hell break loose.. will do that later.

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/brokeroperationexecutor/BrokerOperationExecutorTests.java:
```diff
@@ -0,0 +1,300 @@
+        };
+    }
+
+    private IIpcStrategy getStrategyWithCorruptedResult() {
```
added ServiceException and UserCancelException. (the other two main exceptions that inherited from BaseException - the executor can only throw BaseException)

rpdome commented on common/src/test/java/com/microsoft/identity/common/internal/brokeroperationexecutor/BrokerOperationExecutorTests.java:
```diff
@@ -0,0 +1,300 @@
+        expectBindFailureException(strategyList);
+    }
+
+    // Providing 2 strategies and both succeed.
```
yes.

rpdome commented:
Iteration 2:
- Added IIpcStrategy.Type to IIpcStrategy
  - BrokerCommunicationException now takes this as an argument to identify what IIpcStrategy triggered the exception.
- Extracted the implementation testXXXX() from IpcStrategyTest() to base class.
   - Operation supported by each strategy will eventually be branching out - so it doesn't make sense to have a shared implementation in one place.
        - Broker API will be added to ContentProvider only.
        - AccountManager updateCredential does not support MSAL operation.

AdamBJohnsonx commented on common/src/test/java/com/microsoft/identity/common/internal/ipc/mock/ShadowBoundServiceClientConnectionFailed.java:
```diff
@@ -0,0 +1,43 @@
+public class ShadowBoundServiceClientConnectionFailed<T extends IInterface> {
+    protected @NonNull T connect(@NonNull final String targetServicePackageName)
+            throws ClientException, InterruptedException, TimeoutException, ExecutionException {
+        throw new TimeoutException("Connection timed out");
```
We're shadowing *our own class* here.  There's probably a better way.

AdamBJohnsonx reviewed (approved):
I'm still thinking we're over reliant on Shadow.  We'll have to tackle it eventually, though.

rpdome commented:
force pushed to rebase the branch.
